### PR TITLE
Add new primitives for timeseries preprocessing and anomaly detection

### LINKED ISSUE #111: Improve performance of timeseries and timeseries_errors primitives
Timeseries and timeseries_errors are not efficient enough and should be re-implemented in a way that speeds up their execution.

Also rename `timeseries` to `timeseries_preprocessing` and `timeseries_errors` to `timeseries_anomalies`.

### PR DESCRIPTION
**Pull request by csala**
Resolve #111 

Add new timeseries preprocessing and anomaly detection primitives following the approach explained in this paper https://arxiv.org/abs/1802.04431

### CODE REVIEW

File: mlprimitives/custom/timeseries_anomalies.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    Lastly, we compute a score proportional to the maximum error in the
+    sequence, and finally return the index pairs that correspond to
+    each sequence, along with its score.
+    """
+
+    mean = errors.mean()
+    std = errors.std()
+
+    threshold = find_threshold(errors, z_range[0], z_range[1])
```
Comment (ManuelAlvarezC) [inline mlprimitives/custom/timeseries_anomalies.py]: Here we are splitting `z_range`, to pass it to `find_threshold`, where only be used to generate a range, couldn't just keep it all together and call range unpacking z_range,like `range(*z_range)` on `find_threshold`?

File: mlprimitives/custom/timeseries_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    if isinstance(X, np.ndarray):
+        X = pd.DataFrame(X)
+
+    X = X.sort_values(time_column).set_index(time_column)
+
+    start_ts = X.index.values[0]
+    max_ts = X.index.values[-1]
+
+    values = list()
+    index = list()
```
Comment (ManuelAlvarezC) [inline mlprimitives/custom/timeseries_preprocessing.py]: Here we are creating lists with `list()` and on `rolling_window_sequences` with `[]`would it better if we use just one of the two?

File: mlprimitives/custom/timeseries_anomalies.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        y (array): Ground truth.
+        y_hat (array): Predictions array.
+        smoothing_window (float): Size of the smoothing window, expressed as a proportion
+            of the total length of y.
+        smooth (bool): whether the returned errors should be smoothed with EWMA.
+
+    Returns:
+        (array): errors
+    """
+    errors = [abs(y_h - y_t) for y_h, y_t in zip(y_hat, y)]
```
[the remark this answers is not in the document]
Comment (csala) [reply mlprimitives/custom/timeseries_anomalies.py]: > The two items in the list comprehension have the same name and the second one 'shadows' the first.
> 
> ```python
> >>> a = list(range(3))
> >>> b = a[::-1]
> >>> [(i,i) for i,i in zip(a,b)]
> [(2, 2), (1, 1), (0, 0)]
> ```

No, they aren't! The names used are `y_h` and `y_t`!

But definitely `np.abs` is the way to go, here. I'll change that.

File: mlprimitives/custom/timeseries_anomalies.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    std = errors.std()
+
+    candidates = list()
+    for z in range(min_z, max_z):
+        best = fmin(z_goodness, z, args=(errors, mean, std), disp=False)
+        candidates.append(best[0])
+
+    best_z = candidates[0]
+    best_goodness = np.inf
+    for z in candidates:
```
[the remark this answers is not in the document]
Comment (csala) [reply mlprimitives/custom/timeseries_anomalies.py]: I didn't know about this option! Apparently it is not enabled by default, but you can make `fmin` return the computed value by setting `full_output=True`.
Thanks for the tip!

File: mlprimitives/jsons/mlprimitives.custom.timeseries_anomalies.find_anomalies.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    "modalities": ["timeseries"],
+    "primitive": "mlprimitives.custom.timeseries_anomalies.find_anomalies",
+    "produce": {
+        "args": [
+            {
+                "name": "errors",
+                "type": "ndarray"
+            },
+            {
+                "name": "index",
```
[the remark this answers is not in the document]
Comment (csala) [reply mlprimitives/jsons/mlprimitives.custom.timeseries_anomalies.find_anomalies.json]: Thanks for the suggestion @itinawi , but I think that `index` is more accurate in this case, as here we are dealing with a time index but in other scenarios we might be dealing with numeric indexes.

**Pull request merged**